[Bug] fix high IO after sudden filebeat stop (#35893)

emmanueltouzery opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## Proposed commit message

In case of corrupted log file (which has good chances to happen in case of sudden unclean system shutdown), we set a flag which causes us to checkpoint immediately, but never do anything else besides that. This causes filebeat to just checkpoint on each log operation (therefore causing a high IO load on the server and also causing filebeat to fall behind).

This change resets the logInvalid flag after a successful checkpointing.


## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

I have not in fact tested the PR. I have only checked that it builds. I'm also not 100% sure of the correctness, but the change is really simple and I hope that a maintainer can quickly confirm whether the fix is correct. In any case the current code clearly will cause the issues I described, since `logInvalid` is only ever set to `true`, never to `false`. Therefore I think that the beat trivially cannot recover.

## Related issues

- Closes #35893

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->
See details in #35893 but the issue we fix otherwise manifests itself with a message of:
> WARN memlog/store.go:130 Incomplete or corrupted log file in /var/lib/filebeat/registry/filebeat. Continue with last known complete and consistent state. Reason: unexpected EOF

(or `invalid character '\x00'` instead of EOF)

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > Either there was a build timeout or someone aborted the build.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-39392/detail/PR-39392/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-39392/detail/PR-39392/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-39392/detail/PR-39392/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-39392/detail/PR-39392/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2024-05-03T09:24:19.435Z&rangeTo=2024-05-03T09:44:19.435Z&transactionName=BUILD+Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=c16da3d6de845989e4af9774957a75b9&transactionId=af1f1165a493fdc8)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats





* Duration: 5 min 42 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-39392/detail/PR-39392/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Load a resource file from a library`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-39392/runs/1/steps/82/log/?start=0">here</a></li>
<li>Description: <code>approval-list/elastic/beats.yml</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-39392/runs/1/steps/90/log/?start=0">here</a></li>
<li>Description: <code>githubApiCall: The REST API call https://api.github.com/orgs/elastic/members/emmanueltouzery return the message : java.lang.Exception: httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/emmanueltouzery : httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/emmanueltouzery : Code: 404Error: {"message":"User does not exist or is not a member of the organization","documentation_url":"https://docs.github.com/rest/orgs/members#check-organization-membership-for-a-user"}</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane (Team:Elastic-Agent-Data-Plane)

belimawr commented:
Hi @emmanueltouzery, looking at the code and the original issue, I can see it indeed is a bug. Restarting Filebeat should fix the problem, but I didn't have time to test it.

To get this merged we'd need some tests, setting `logInvalid` to `false` can be done right after truncating the log file.

Would you be interested in further working on this PR? I didn't take a deep look at our current tests, but I believe it would be simple to write a unit test that simulates this condition and ensure the fix works as expected.

I can try guiding you on this.

emmanueltouzery commented:
Well first off I'm delighted that the issue will likely be fixed

emmanueltouzery commented:
Sorry for closing and reopening, I'm typing on a smartphone.

I can confirm that restarting filebeat fixes the issue.

So if you think it makes sense for you, I can try to improve the PR further with your guidance and worst-case you take over. 

To start with, could you tell me in which file the test should be added?

With regards to when to reset the flag, you mention truncating the log, is that the checkpointing operation or does that happen elsewhere?

emmanueltouzery commented:
ping?

emmanueltouzery commented:
@belimawr I'm interested in trying to get a fix merged for this issue. 

could you tell me in which file the test should be added?

With regards to when to reset the flag, you mention truncating the log, is that the checkpointing operation or does that happen elsewhere?

Thank you!

belimawr commented:
Hey @emmanueltouzery, sorry the delay in replying.

> With regards to when to reset the flag, you mention truncating the log, is that the checkpointing operation or does that happen elsewhere?

Yes, the checkpoint operation, once the log file is truncated, it will be in a valid state, so we can already reset the `logInvalid` flag.

https://github.com/elastic/beats/blob/23d106637069322b39b7f808b3940a788c82660c/libbeat/statestore/backend/memlog/diskstore.go#L396-L417

> could you tell me in which file the test should be added?

As you're modifying `diskstore.go`, you can create a test file next to it, `diskstore_test.go` and add at least one unit test there. I believe it should be simple to instantiate the disk store pointing to a file you can easily corrupt, validate that `logInvalid` gets reset and new data starts to be added to the log file.

Ping me if  you need more guidance.  The other tests in this package might help you.

emmanueltouzery commented:
I have updated the PR now. I added the test against store, not diskstore, because the read validation is done in store. I also reused an invalid truncated log sample file from another test. Let me know whether that looks good!

I saw the CI fail, the errors seem unrelated to my code. Let me know if I'm wrong.

emmanueltouzery commented:
I also added a second commit to fix deprecation warnings in diskstore.

belimawr commented on libbeat/statestore/backend/memlog/store_test.go:
```diff
@@ -0,0 +1,54 @@
+	store, err := openStore(logp.NewLogger("test"), path, 0660, 4096, false, func(_ uint64) bool {
+		return false
+	})
+	assert.NoError(t, err)
+
+	assert.Equal(t, true, store.disk.logInvalid)
+
+	err = store.logOperation(&opSet{K: "key", V: mapstr.M{
+		"field": 42,
+	}})
+	assert.NoError(t, err)
+	assert.Equal(t, false, store.disk.logInvalid)
```
A few things here:
1. You can use `require` instead of assert when a condition that won't let the test continue happens, this way the test is interrupted
2. It's good practice to always add a message explaining why the test failed.

```suggestion
	require.NoError(t, err, "openStore must succeed")
	require.True(t, store.disk.logInvalid, "expecting the log file to be invalid")

	err = store.logOperation(&opSet{K: "key", V: mapstr.M{
		"field": 42,
	}})
	require.NoError(t, err, "logOperation must succeed")
	require.Equal(t, false, store.disk.logInvalid, "log file must be valid")
```

There are a couple of other things that would be good to assert here:
1. The checkpoint file was created (check the file `7.json` was created)
2. The log file is not corrupted any more/was truncated (check `log.json` is emtpy/0 bytes in size).

belimawr reviewed (changes requested):
The PR is looking great @emmanueltouzery! There are a few things to improve/add and we'll be ready to merge it :star_struck:

emmanueltouzery commented:
updated the PR, thanks for the great feedback!

now the docs fail, apparently i should run `make update`. if I run that in the libbeat subfolder, it succeeds but doesn't touch the disk. if I run it in the root of the folder, it fails with that output:
```
Generating NOTICE
go mod tidy
go: github.com/prometheus/prometheus@v1.8.2-0.20210701133801-b0944590a1c9 requires
        github.com/influxdata/influxdb@v1.9.2 requires
        github.com/influxdata/flux@v0.113.0 requires
        github.com/SAP/go-hdb@v0.14.1: reading github.com/SAP/go-hdb/go.mod at revision v0.14.1: unknown revision v0.14.1
make: *** [Makefile:172: notice] Error 1
```

so i'm not sure how to fix that.

belimawr commented:
> updated the PR, thanks for the great feedback!
> 
> now the docs fail, apparently i should run `make update`. if I run that in the libbeat subfolder, it succeeds but doesn't touch the disk. if I run it in the root of the folder, it fails with that output:
> 
> ```
> Generating NOTICE
> go mod tidy
> go: github.com/prometheus/prometheus@v1.8.2-0.20210701133801-b0944590a1c9 requires
>         github.com/influxdata/influxdb@v1.9.2 requires
>         github.com/influxdata/flux@v0.113.0 requires
>         github.com/SAP/go-hdb@v0.14.1: reading github.com/SAP/go-hdb/go.mod at revision v0.14.1: unknown revision v0.14.1
> make: *** [Makefile:172: notice] Error 1
> ```
> 
> so i'm not sure how to fix that.

That's odd.

I tried here with your branch and it is working for me, here is what you can try:
1. Ensure all modified files are commited/stashed
2. Ensure you're running Go 1.21.10 and your environment has the $PATH correctly set for the `go` command and binaries `go install` might install.
3. Make sure you have [`mage`](https://magefile.org/) installed
4. From the root of the repository run:
    ```
    make clean
    make update
    make check
    ```

belimawr commented:
I was looking again into it, and what is failing is one of the files is not formated as expected, to fix it:
```
cd libbeat
mage fmt
```
Then commit the updated file.

emmanueltouzery commented:
thanks for the help, indeed that did it :tada:

belimawr commented on libbeat/statestore/backend/memlog/store_test.go:
```diff
@@ -0,0 +1,55 @@
+		"field": 42,
+	}})
+	require.NoError(t, err, "logOperation must succeed")
+	require.Equal(t, false, store.disk.logInvalid, "log file must be valid")
```
Sorry I missed that in my last review, but here you can use `require.False`
```suggestion
	require.False(t, store.disk.logInvalid, "log file must be valid")
```

belimawr reviewed (changes requested):
The PR is looking great! Just one small change and I'll approve it.

emmanueltouzery commented:
applied your suggestion :+1: 
i was wondering, will this get backported to 7.x as well?

dliappis commented:
test

belimawr commented:
> applied your suggestion 👍 i was wondering, will this get backported to 7.x as well?

I didn't look at the 7.17 code, but if the bug is also there, yes, we can backport it. I'll add the label to create the backport PR

belimawr commented:
The lastest test run failed because of a known issue: https://github.com/elastic/beats/issues/39698. We'll have to wait for the test to be fixed or skipped and then update this PR with `main`.

emmanueltouzery commented:
I can confirm that 7.x is affected too.

emmanueltouzery commented:
> The lastest test run failed because of a known issue: https://github.com/elastic/beats/issues/39698. We'll have to wait for the test to be fixed or skipped and then update this PR with main.

but i see other things getting merged despite this? and no activity on #39698 in the last five days?

i'm mentioning this because from my point of view this issue we're fixing here is not that benign... i'm hoping the fix goes in sometime soon.

belimawr commented:
> i'm mentioning this because from my point of view this issue we're fixing here is not that benign... i'm hoping the fix goes in sometime soon.

You're right, the issue is not with your PR. We do our best not to force merge PRs if CI is failing, even if it is for unrelated reasons. If there is a flaky tests, some times not all PRs fails on it, so they get merged.

I'll update your branch with the latest `main` and re-trigger CI.

dliappis commented:
run docs-build

belimawr commented:
Thank you so much for the help @emmanueltouzery !!

emmanueltouzery commented:
I see this is not in 8.14.0. so this will be released in 8.15.0?

Thank you, and great we got this in!

pierrehilbert commented:
Yes this will land in 8.15.0 in the current state but I think we should backport it to have it in 8.14.1.
@belimawr WDYT?
